src/olympia/yara/models.py

```diff
@@ -28,4 +28,9 @@ def add_match(self, rule, tags=None, meta=None):
 
     @property
     def matched_rules(self):
-        return [match['rule'] for match in self.matches]
+        matched_rules = []
+        for match in self.matches:
+            rule = match['rule']
+            if rule not in matched_rules:
+                matched_rules.append(rule)
+        return matched_rules
```

Comment:
return `list({match['rule'] for match in self.matches})`

Reply:
Will this keep the order of the items in the list? I am not sure I need this but when I used `list(set())`, another test case failed because `rule1` came after `rule2`.

Reply:
It won't.  Though, thinking about it, if having duplicates is a problem why not check uniqueness in `add_match` instead?

Reply:
We should store all the occurrences in database because it will give hints about which rule matched on which file. This computed property is used in the list view of the Django admin.

Maybe it does not matter if the order is not preserved?

Reply:
I wouldn't have thought it would matter, except for readability in django admin, though I don't have full sight of the whole project so 🤷‍♂

Reply:
This property is used in the right column here:

![](https://user-images.githubusercontent.com/217628/62315496-b4ebf480-b495-11e9-959f-78c103278cad.png)

I suppose if we don't preserve the order and a user refreshes the list many times, then the values in this column will likely change, but that does not look like a huge problem... So maybe we should keep things simple?

Reply:
You could return fully `sorted` rules every time to make the order stable ?

Reply:
Yep, that would make sense, I guess. So `list(sorted(set()))`?

Reply:
`sorted({match['rule'] for match in self.matches})` should be enough - `sorted` returns a list from an iterable.